[LLDB] Update DIL to handle smart pointers; add more tests.

cmtice opened the pull request:
This updates the DIL implementation to handle smart pointers (accessing field members and dereferencing) in the same way the current 'frame variable' implementation does.  It also adds tests for handling smart pointers, as well as some additional DIL tests.

cmtice committed:
[LLDB] Update DIL to handle smart pointers; add more tests.

This updates the DIL implementation to handle smart pointers (accessing
field members and dereferencing) in the same way the current 'frame
variable' implementation does.  It also adds tests for handling smart pointers,
as well as some additional DIL tests.

labath reviewed (commented):
The code looks good, but it seems that you need to reformat it (both c++ and python).

Instead of hardcoding libc++ implementation details, it would be better to write your own "smart" pointer. libc++ changes the internal representation fairly frequently, and we shouldn't rely on it where it isn't necessary.  If this test is meant to stay here ~forever, I think that's what we should do, but I'm not sure what is the long-term future of these tests.

Have you thought about what happens to them when we flip the flag (and maybe after the old implementation is removed)? I would expect that most of this functionality is already tested elsewhere, so I have a feeling that some sort of deduplication would be in order. And since you've found this issue by running the test suite, I assume that this is tested elsewhere (maybe inside the std::shared_ptr formatter test?)..

cmtice committed:
Remove tests that depend on internals of smart ptr implementation.

cmtice committed:
Fix clang-format issues with tests.

cmtice committed:
Add fake smart pointer test.

Test synthetic format provider similar to LibCxx smart pointers,
but independent of actual smart pointer implementation (so not brittle test).

cmtice commented:
I *think* I have addressed all the review comments; please take another look.

cmtice committed:
Fix clang-format issues.

cmtice committed:
Fix another clang-format issue.

labath reviewed (commented):
The code looks good. I just have issues with the test(s). Please, see inline comments.

cmtice committed:
Address reviewer comments:

- Moved	DIL SharedPtr and UniquePtr tests to
 API/functionalities/data-formatter/data-formatter-stl/libcxx/dil_shared_ptr
and
 API/functionalities/data-formatter/data-formatter-stl/libcxx/dil_unique_ptr
- Moved	my new test to API/command/frame/var-dil/basics/SyntheticDereference.
- Changed 'smart_ptr' to 'wrap_ptr' in new test.
- Changed internal field name from '__ptr_' to 'ptr.
- Remove dead code from	test.
- Update pointer dereference test to test some expected	vs actual addresses,
  as suggested.

cmtice commented on lldb/test/API/commands/frame/var-dil/basics/PointerDereference/TestFrameVarDILPointerDereference.py:
```diff
@@ -0,0 +1,34 @@
+        )
+
+        self.runCmd("settings set target.experimental.use-DIL true")
+        self.expect_var_path("*p_int0", value="0")
+        self.expect_var_path("*cp_int5", value="5")
+        self.expect_var_path("&pp_void0[2]", type="void **")
+        self.expect_var_path("**pp_int0", value="0")
+        self.expect_var_path("&**pp_int0", type="int *")
```
Thanks for the suggestion! Done.

cmtice committed:
Fix clang-format issues.

labath reviewed (approved):
Thanks. Could you also merge them into the existing tests (as a separate test method)? The existing tests are of the same type ("break and inspect a bunch of variables") so it should be easy enough and this way one can see all tests for a type in single place.

cmtice committed:
Move DIL tests for shared & unique pointers into existing dataformatter tests.

cmtice commented:
> Thanks. Could you also merge them into the existing tests (as a separate test method)? The existing tests are of the same type ("break and inspect a bunch of variables") so it should be easy enough and this way one can see all tests for a type in single place.

Done? I think?

labath reviewed (commented):
I meant adding a new test method (`def test_something(self): self.build(); self.run(); your code`), but this is fine too, especially if the switch to the new implementation is imminent.
